Readme.md
This is a static site generator that generates web versions of CWRU career fair employer guides.


    > python3 careerfair.py --help
    usage: careerfair.py [-h] {spring,fall} year data

    Generate static HTML site for the CWRU career fair employer guide

    positional arguments:
      {spring,fall}
      year
      data           CSV file

    optional arguments:
      -h, --help     show this help message and exit

Output will be in `{season}{year}/`. `index.html` in the root directory will link to files in `{season}{year}/` so that Github Pages shows the index as the main page.

Files in `content/` will be copied to the output directory.

The CSV columns are defined in `company.py`.

Make sure changes get to the `gh-pages` branch and are pushed from there.


fall2011/History
<html>
    <head>
        <link rel="stylesheet" type="text/css"  
            href="blueprintcss/typography.css"/>
        <link rel="stylesheet" type="text/css"  
            href="blueprintcss/grid.css"/>
        <title>%s</title>
        <script type="text/javascript">

          var _gaq = _gaq || [];
          _gaq.push(['_setAccount', 'UA-25939700-1']);
          _gaq.push(['_trackPageview']);

          (function() {
            var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
            ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ ga.js';
            var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
            })();

        </script>
    </head>
    <body>
    <div class="container" id="main">
        <a href="http://studentaffairs.case.edu/careers/"><img src="banner.jpg" style="float: right"/></a>
        <h1 id="title">Companies looking for History majors</h1>
        <h3>Produced by members of the <a href="http://goto.case.edu/">CWRU Hacker Society</a></h3>
        <div class="span-8 colborder" id="sidebar">
            <h2>Majors</h2>
            
        </div>
        <div class="span-15 last" id="content">
            
                <h2><a href="http://datel-group.com">DATEL Software Solutions</a></h2>
                <strong>Degrees:</strong> Bachelors<br/>
                <strong>Position types:</strong> Part-Time, Full Time, Practicum, Co-op, Internship<br/>
                <strong>Majors:</strong> Biology/Systems Biology/Evolutionary Biology, Physics/Astronomy, Sociology, Psychology, Geology/Environmental Studies, Biochemistry, Interdisciplinary Programs and Modern Languages, Engineering Mechanics (G), All Sciences, Fluid and Thermal (G), Chemical, Electrical, Biomedical, Political Science, All Weatherhead, Finance (G), Macromolecular Science (G), Management -- Finance (U), Nutrition, English/Classics/Literature, Mathematics/Statistics, Supply Chain (G), Anthropology, Management -- Marketing (U), Systems & Control, Computer Science, Aerospace, All Engineering, Positive Organizational Development (G), Economics, MBA (G), Accounting/Accountancy, All Arts, Humanities, & Social Sciences, Operations Research (G), Engineering Physics (U), Communication Disorders/Cognitive Science, Bachelors, Master of Engineering Management (G), Mechanical, Philosophy/Religious Studies, Chemistry/Natural Sciences, History, Music/Music Education/Theatre/Dance, Civil, Materials Science, Teacher Education, Computer Engineering, Polymer Science, Art History/Art Education, Organizational Behavior (G)<br/>
                <strong>Locations:</strong> Pittsburgh, PA<br/>
                <strong>F1:</strong> No<br/>
                <br/>
                <strong>Contact:</strong> <a href="mailto:dlemke@datel-group.com">Dan Lemke</a>, Principal<br/>
                <strong>Phone:</strong> 724-940-5150
                    <strong>Fax:</strong> 724-940-5147<br/>
                <strong>Address:</strong> DATEL Software Solutions12300 Perry HighwaySte. 306Wexford, PA. 15090<br/>
            
                <h2><a href="http://newyorklife.com/corporatecareers">New York Life Insurance Company</a></h2>
                <strong>Degrees:</strong> Bachelors<br/>
                <strong>Position types:</strong> Part-Time, Full Time<br/>
                <strong>Majors:</strong> Sociology, Psychology, Management -- Marketing (U), All Arts, Humanities, & Social Sciences, Economics, Bachelors, Political Science, Finance (G), Management -- Finance (U), History<br/>
                <strong>Locations:</strong> Cleveland, OH<br/>
                <strong>F1:</strong> No<br/>
                <br/>
                <strong>Contact:</strong> <a href="mailto:nancy_maksomski@newyorklife.com">Nancy Maksomski</a>, Dirctor Talent Acquisition<br/>
                <strong>Phone:</strong> 212-576-5714
                    <strong>Fax:</strong> 212-447-4292<br/>
                <strong>Address:</strong> 51 Madison AvenueRoom 108NY, NY 10013<br/>
            
        </div>
    </body>
</html>

company.py
from collections import namedtuple
from functools import reduce
import itertools
import re


Cols = namedtuple( 'Cols', [
    'company_name',
    'url',
    'description',
    'junk',
    'contact_name',
    'contact_title',
    'address',
    'phone',
    'fax',
    'contact_email',
    'position_types',
    'degrees',
    'majors_artsci',
    'majors_eng',
    'majors_mgmt',
    'majors_professional',
    'f1',
    'locations'])

def parse_nested_lists(mlists):
    """
    ['abc;def', 'ghi;jkl'] -> ['abc', 'def', 'ghi', 'jkl']
    """
    for mlist in mlists:
        for m in mlist.split(';'):
            if m.strip():
                yield m.strip()

def liberate_semicolon_strings(s):
    return [m.strip() for m in s.split(';') if m.strip()]

def html_escape(text):
    text = text.replace('&', '&amp;')
    text = text.replace('"', '&quot;')
    text = text.replace("'", '&#39;')
    text = text.replace(">", '&gt;')
    text = text.replace("<", '&lt;')
    text = text.replace("’", '&apos;')
    text = text.replace("–", '&#150;')
    return text

LOWER = 0
UPPER = 1
DIGIT = 2
WHITESPACE = None

def fix_address(a):
    # Fix the horribly broken nonexistent line breaks from the CSV
    words = [""]
    tok_type = WHITESPACE
    for c in a:
        new_space = False
        echo = False
        if c.islower():
            echo = True
            tok_type = LOWER
        elif c.isupper():
            echo = True
            if tok_type != UPPER and tok_type != WHITESPACE and words[0]:
                new_space = True
            tok_type = UPPER
        elif c.isdigit():
            echo = True
            if tok_type != DIGIT and tok_type != WHITESPACE and words[0]:
                new_space = True
            tok_type = DIGIT
        elif c.isspace():
            if tok_type != WHITESPACE:
                new_space = True
            tok_type = WHITESPACE
        elif c != ',':
            echo = True
        if new_space:
            words.append("")
        if echo:
            words[-1] = words[-1] + c
    words = [w.strip() for w in words]
    # OKAY, now let's try to do something with that mess
    state, zipcode = words[-2:]
    city = words[-3]

    if city in ('York',):
        city = 'New York'
        words[-4] = words[-4][:-3]

    if len(words) <= 4 or city == 'Ave.':
        return ' '.join(words) + '\n'

    index_of_first_number = 0
    while not words[index_of_first_number].isdigit() and not words[index_of_first_number] == 'P.O.':
        index_of_first_number += 1
    if index_of_first_number >= len(words)-3 or index_of_first_number <= 1:
        index_of_first_number = 0
    if index_of_first_number == 0:
        return "%s<br/>%s, %s %s" % (' '.join(words[:-3]), city, state, zipcode)
    else:
        firstline = ' '.join(words[:index_of_first_number])
        secondline = ' '.join(words[index_of_first_number:-3])
        return "%s<br/>%s<br/>%s, %s %s" % (firstline, secondline, city, state, zipcode)


class Company(object):
    def __init__(self, *values):
        cols = Cols(*values)
        self.name = cols.company_name
        self.url = cols.url
        self.description = html_escape(cols.description)
        self.contact_name = cols.contact_name
        self.contact_title = cols.contact_title
        self.contact_email = cols.contact_email
        # self.address = cols.address
        self.address = fix_address(cols.address)
        self.phone = cols.phone
        self.fax = cols.fax
        self.position_types = liberate_semicolon_strings(cols.position_types)
        self.degrees = liberate_semicolon_strings(cols.degrees)
        
        semi_mlists = (cols.majors_artsci, cols.majors_eng,
                       cols.majors_mgmt, cols.majors_professional)
        self.mlists = [liberate_semicolon_strings(x) for x in semi_mlists]
        self.majors = reduce(lambda a, b: set(a) | set(b), self.mlists, set())
        
        self.f1 = (cols.f1.lower()  == 'y')
        self.locations = cols.locations
        self.session = ''
        self.oci = False
        self.ocif = False
    
    def shortstring(self):
        return "%s - %s" % (self.name, self.url)
    
    def __str__(self):
        title = "%s - %s" % (self.name, self.url)
        titledashes = ''.join(['=' for c in title])
        contact = "%s (%s, %s)" % (self.contact_name, self.contact_email, self.phone)
        
        return '\n'.join([titledashes, 
                          title, 
                          titledashes, 
                          contact, 
                          self.locations, 
                          ', '.join(self.majors), 
                          '', 
                          self.description, 
                          '', 
                          self.session])
    


careerfair.py
import argparse
import csv
import itertools
import os
import re
import shutil

import jinja2

from company import *


def sanitized(v):
    return re.sub('[^a-zA-Z_]', '', v)


def commajoin(v):
    return ', '.join(v)


def yesorno(v):
    return 'Yes' if v else 'No'


def render(name, out, *args, **kwargs):
    env = jinja2.Environment(loader=jinja2.FileSystemLoader('.'))
    env.filters['sanitized'] = sanitized
    env.filters['commajoin'] = commajoin
    env.filters['yesorno'] = yesorno
    t = env.get_template(name)
    with open(out, 'w') as f:
        f.write(t.render(*args, **kwargs))


def make_arg_parser():
    parser = argparse.ArgumentParser(description="Generate static HTML site for the CWRU career fair employer guide")
    parser.add_argument('season', choices=('spring', 'fall'))
    parser.add_argument('year', type=int)
    parser.add_argument('data', help="CSV file")
    return parser



def process(args):
    # 1. Convert

    # CSV -> sorted list of all companies
    with open(args.data, 'r') as csv_f:
        reader = csv.reader(csv_f, dialect=csv.excel)
        sorted_companies = sorted((Company(*row) for row in list(reader)[1:]),
                                  key=lambda c: c.name.lower())

    # 2. Store sets of values

    # The CSV breaks majors into 4 categories, so we do too
    mlists = [set() for i in range(4)]

    # But we also combine them into one master set
    majors = set()

    # also store a set of all degrees
    all_degrees = set()

    for c in sorted_companies:
        # set unions
        majors = majors | set(c.majors)
        all_degrees = all_degrees | set(c.degrees)
        for i in range(len(mlists)):
            mlists[i] = mlists[i] | set(c.mlists[i])

    # 3. Construct sidebar sections [(section_title, list_items)]

    titles = ('Degrees',
              'College of Arts and Sciences',
              'Case School of Engineering',
              'Weatherhead School of Management',
              'Professional Schools')
    sections = list(zip(titles, [sorted(all_degrees)] + [sorted(l) for l in mlists]))

    # 4. Initialize output directory
    out_dir = '{args.season}{args.year}'.format(args=args)
    shutil.rmtree(out_dir)

    os.mkdir(out_dir)
    shutil.copytree('content', os.path.join(out_dir, 'content'))

    # 5. Generate output files

    season_name = {'fall': 'Fall', 'spring': 'Spring'}[args.season]
    title = 'CWRU Career Fair {0} {1} Employer Guide'.format(season_name, args.year)
    companies = sorted_companies
    render('template.html', os.path.join(out_dir, 'index.html'), **locals())
    render('template.html', 'index.html', prefix='{0}/'.format(out_dir),**locals())

    for major in majors:
        title = 'Companies looking for {0} majors'.format(major)
        companies = [c for c in sorted_companies if major in c.majors]
        render('template.html',
               os.path.join(out_dir, sanitized(major) + '.html'),
               **locals())
    for degree in all_degrees:
        title = 'Companies looking for {0} students'.format(degree)
        companies = [c for c in sorted_companies if degree in c.degrees]
        render('template.html',
               os.path.join(out_dir, sanitized(degree) + '.html'),
               **locals())

if __name__ == '__main__':
    parser = make_arg_parser()
    process(parser.parse_args())


gen.py
# -*- coding: utf-8 -*-
import codecs

header = u"""
<html><head>
    <link rel="stylesheet" type="text/css"  
        href="blueprintcss/typography.css"/>
    <link rel="stylesheet" type="text/css"  
        href="blueprintcss/grid.css"/>
    <title>%s</title>
    <script type="text/javascript">

      var _gaq = _gaq || [];
      _gaq.push(['_setAccount', 'UA-4517625-4']);
      _gaq.push(['_trackPageview']);

      (function() {
        var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
        ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
        var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
      })();

    </script>
</head><body>
    <div class="container" id="main">
        <a href="http://studentaffairs.case.edu/careers/"><img src="banner.jpg" style="float: right"/></a>
        <h1 id="title">%s</h1>
        <h3>Produced by members of the <a href="http://goto.case.edu/">CWRU Hacker Society</a></h3>
        <div class="span-8 colborder" id="sidebar">
"""

footer = u"""</div></div></body></html>"""

sidebar_divider = u"""</div><div class="span-15 last" id="content">"""

sanitize_major = lambda major: major.replace('/', '_')

def html_escape(text):
    text = text.replace(u'&', u'&amp;')
    text = text.replace(u'"', u'&quot;')
    text = text.replace(u"'", u'&#39;')
    text = text.replace(u">", u'&gt;')
    text = text.replace(u"<", u'&lt;')
    text = text.replace(u"’", u'&apos;')
    text = text.replace(u"–", u'&#150;')
    return text

def write_majors(f, mlists):
    f.write("""<h2>Majors</h2>""")
    titles = ('', 'College of Arts and Sciences', 
              'Case School of Engineering', 'Weatherhead School of Management', 
              'Professional Schools')
    for l, t in zip(mlists, titles):
        if t:
            f.write("""<h3>%s</h3>""" % t)
        for m in sorted(l):
            f.write('<a href="%s.html">%s</a><br/>' % (sanitize_major(m), m))
    f.write(sidebar_divider)

def write_company(f, c):
    f.write(u'<h2>%s</h2>' % c.link())
    f.write(u'<strong>Degrees:</strong> %s<br/>' % ', '.join(c.degrees))
    f.write(u'<strong>Position types:</strong> %s<br/>' % ', '.join(c.position_types))
    f.write(u'<strong>Majors:</strong> %s<br/>' % ', '.join(c.majors))
    f.write(u'<strong>Locations:</strong> %s<br/>' % c.locations)
    f.write(u'<strong>F1:</strong> %s<br/>' % (u'Yes' if c.f1 else u'No'))
    f.write(u'<br/>')
    
    f.write(u'<strong>Contact:</strong> <a href="mailto:%s">%s</a>, %s<br/>' % (c.contact_email, c.contact_name, c.contact_title))
    f.write(u'<strong>Phone:</strong> %s <strong>Fax:</strong> %s<br/>' % (c.phone, c.fax))
    f.write(u'<strong>Address:</strong> <br/>%s<br/>' % c.address)
    f.write(u'<br/>')
    
    if c.oci:
        f.write(u'On-Campus Interviews<br/>')
    if c.ocif:
        f.write(u'On-Campus Interview Friday<br/>')
    if c.session:
        f.write('u<strong>Session:</strong> %s<br/>' % c.session)
    if c.oci or c.ocif or c.session:
        f.write('u<br/>')
    
    f.write(html_escape(u'<br/>'.join(c.description.splitlines())))

def gen_index(majors, companies, path, title):
    with codecs.open('%s/index.html' % path, mode='w', encoding='utf-8') as f:
        f.write(header % (title, title))
        write_majors(f, majors)
        for c in companies:
            write_company(f, c)
        f.write(footer)

def gen(major, majors, companies, path):
    with codecs.open('%s/%s.html' % (path, sanitize_major(major)), mode='w', encoding='utf-8') as f:
        t = u'Companies looking for %s majors' % major
        f.write(header % (t, t))
        write_majors(f, majors)
        for c in [x for x in companies if major in x.majors]:
            write_company(f, c)
        f.write(footer)
